Tutorial - LOJ 1089 (en) (#450)
lightoj-dev/problem-tutorials: 1089/en.md

# LOJ 1089 - Points in Segments (II)

## Summary
The problem involves a set of one-dimensional segments and a set of points. The objective is to count, for each point, how many segments include that point within their range. 

Note: **AA** is a valid segment.

## Solution
First let's simplify the problem a bit. Suppose there is no such segment like: **AA** 

Let's think of each segment **AB** as of two points where **A** is the opening of the segment and **B** the ending of it. 

> Consider the set **S** of all opening and ending points of the segments sorted in increasing order. Iterating over the set from the begining, whenever we find an opening point **S[i]**, we increment the **counter** (starting with 0) by 1, indicating for any point **p** such that **S[i] <= p <= S[i+1]**,  if **S[i+1]** is an ending point, otherwise **S[i] <= p < S[i+1]** will be contained by that many of segments. If we find an ending point **S[i]**, we know for sure we had the opening point of that segment before, so it's time to close the segment, so we decrement the **counter** by 1, now indicating for any point **p** such that **S[i] < p <= S[i+1]**,  if **S[i+1]** is an ending point, otherwise **S[i] < p < S[i+1]** will be contained by that many of segments.

Reviewer: Let's take a step back and see if we can simplify this paragraph. I understand what you are trying to explain here, but I feel there are too many complicated sentences jumbled together. Let's explain it simply.
While doing so, let's also deviate away from any implementation details, for example, not bother explaining how you may write this in code.

Author: Simplified!

We can sort our queries so that we can calculate answers for them in increasing order as we iterate over the set **S**.

> That was the general approach. Now, solving the original problems is all about handling the exception case where we may have the same point for many openings and endings. See the code and documentation for understading how to handle duplicate points.

Reviewer: ~documentation~

comments

Author: Fixed!

## Complexity

> - Time Complexity: O(T * (N $log{_2}{N}$ + Q $log{_2}{Q}$).

Author: How would you suggest to write down the complexity?

> - Time Complexity: O(T * (N $log{_2}{N}$ + Q $log{_2}{Q}$).

Reviewer: Just write it in plain English. Use `lg` instead of `log_2` if you prefer it over `log`. In CS, `log` typically translates to `log_2`.

Let's think of each segment **AB** as of two points where **A** is the opening of the segment and **B** the ending of it. 

> Consider the set of all opening and ending points of the segments sorted in increasing order. Iterating over the set from the beginning, only two case can arise. 

Reviewer: ```suggestion
Consider the set of all opening and ending points of the segments sorted in increasing order. Iterating over the set, there may be two scenarios: 
```

> 1. An opening point **p** indicating the segment (of that opening point **p**) has started running. By means of saying that any point that comes after it (or equal to that) and not after the ending of it will be contained in that segment. 

Reviewer: ```suggestion
1. An opening point `A` of segment `AB` is found. `AB` has started. Any point that comes after `A` (or equal to `A`) until we reach the endpoint `B` will be on that segment.
```

> 2. An ending point **p** indicating an ongoing segment has just been closed now and we can't take any point after that into the account of that segment (whose ending is **p**). 

Reviewer: ```suggestion
2. An ending point `B` of segment `AB` is found. Ongoing segment `AB` is closed. We can't take any point after that into account. 
```
